test/jdk/java/net/httpclient/HttpClientLocalAddrTest.java

```diff
@@ -67,88 +69,191 @@
-            // don't let proxies interfere with the client addresses received on the
-            // HTTP request, by the server side handler used in this test.
-            builder.proxy(HttpClient.Builder.NO_PROXY);
+    /**
+     * Stops the server and returns (instead of throwing) any exception that might
+     * have occurred during stop. If {@code prevException} is not null then any
+     * exception during stop of the {@code server} will be added as a suppressed
+     * exception to the {@code prevException} and the {@code prevException} will be
+     * returned.
+     */
+    private static Exception safeStop(HttpServerAdapters.HttpTestServer server, Exception prevException) {
+        if (server == null) {
+            return null;
+        }
+        var serverAuthority = server.serverAuthority();
+        try {
+            server.stop();
+        } catch (Exception e) {
+            System.err.println("Failed to stop server " + serverAuthority);
+            if (prevException == null) {
+                return e;
+            }
+            prevException.addSuppressed(e);
+            return prevException;
+        }
+        return prevException;
+    }
+
+    @DataProvider(name = "params")
+    private Object[][] paramsProvider() throws Exception {
+        final List<Object[]> clients = new ArrayList();
+        final URI[] requestURIs = new URI[]{httpURI, httpsURI, http2URI, https2URI};
+        final Predicate<URI> requiresSSLContext = (uri) -> uri.getScheme().equals("https");
+        for (var requestURI : requestURIs) {
+            final var configureClientSSL = requiresSSLContext.test(requestURI);
             // no localAddr set
-            clients.add(new Object[]{builder.build(), null});
+            clients.add(new Object[]{
+                    newBuilder(configureClientSSL).build(),
+                    requestURI,
+                    null
+            });
             // null localAddr set
-            clients.add(new Object[]{builder.localAddress(null).build(), null});
+            clients.add(new Object[]{
+                    newBuilder(configureClientSSL).localAddress(null).build(),
+                    requestURI,
+                    null
+            });
             // localAddr set to loopback address
-            var loopbackAddr = InetAddress.getLoopbackAddress();
-            clients.add(new Object[]{builder.localAddress(new InetSocketAddress(loopbackAddr, 0)).build(), loopbackAddr});
+            final var loopbackAddr = InetAddress.getLoopbackAddress();
+            clients.add(new Object[]{
+                    newBuilder(configureClientSSL)
+                            .localAddress(new InetSocketAddress(loopbackAddr, 0))
+                            .build(),
+                    requestURI,
+                    loopbackAddr
+            });
             // anyAddress
             if (Boolean.getBoolean("java.net.preferIPv6Addresses")) {
                 // ipv6 wildcard
```

Comment:
::1 is the ipv6 loopback - not the ipv6 wildcard

Reply:
You may encounter some issues when testing with https and IPv6 - Michael is working on a fix to add the loopback addresses as SNI aliases to the SimpleSSLContext certificate:
https://bugs.openjdk.java.net/browse/JDK-8278312

Reply:
You could also replace `Boolean.getBoolean("java.net.preferIPv6Addresses")` with a call to `IPSupport::preferIPv6Addresses()`

Reply:
> ::1 is the ipv6 loopback - not the ipv6 wildcard

Oops, that's embarrassing and it isn't even a copy/paste mistake. I have been in part of the test so many times while cleaning up/debugging and yet didn't notice this. Fixed it in the latest PR update now.

I'll update the rest of this test with your suggestions shortly. Thank you Daniel for the reviews.



>

Reply:
> You may encounter some issues when testing with https and IPv6 - Michael is working on a fix to add the loopback addresses as SNI aliases to the SimpleSSLContext certificate: https://bugs.openjdk.java.net/browse/JDK-8278312

I had a look at this one. Turns out my test wasn't impacted by it because the test ends up using a hostname based URLs (`localhost`) for requests and the issue only impacts requests that use IP addresses for the requests (where the server then presents a hostname based cert causing the cert validation failure.).